File: spring2013/codeWar/LRS/Clients/PlayerCSharpAI/api/GameMap.cs
// Created by Windward Studios, Inc. (www.windward.net). No copyright claimed - do anything you want with this code.

using System.Drawing;
using System.Linq;
using System.Xml.Linq;

namespace PlayerCSharpAI.api
{
	/// <summary>
	/// The map of the game.
	/// </summary>
	public class GameMap
	{

		/// <summary>
		/// Create the map from the passed across XML
		/// </summary>
		/// <param name="element">The parent map node.</param>
		public GameMap(XElement element)
		{
			int width = int.Parse(element.Attribute("width").Value);
			int height = int.Parse(element.Attribute("height").Value);
			Squares = new MapSquare[width][];
			for (int x = 0; x < width; x++)
			{
				Squares[x] = new MapSquare[height];
				for (int y = 0; y < height; y++)
				{
					XElement elemSquare =
						(from sq in element.Elements("square")
						 where ((int) sq.Attribute("x") == x) && ((int) sq.Attribute("y") == y)
						 select sq).First();
					Squares[x][y] = new MapSquare(elemSquare);
				}
			}
		}

		/// <summary>
		/// The width of the map. Units are squares.
		/// </summary>
		public int Width
		{
			get { return Squares.Length; }
		}

		/// <summary>
		/// The height of the map. Units are squares.
		/// </summary>
		public int Height
		{
			get { return Squares[0].Length; }
		}

		/// <summary>
		/// The map squares. This is in the format [x][y].
		/// </summary>
		public MapSquare [][] Squares { get; protected set; }

		/// <summary>
		/// Location (map squares) of the flags in order.
		/// </summary>
		public Point[] Flags { get; protected set; }

		/// <summary>
		/// Get a specific map square.
		/// </summary>
		/// <param name="location">The x, y point for the square desired.</param>
		/// <returns>The requested square.</returns>
		public MapSquare GetSquare(Point location)
		{
			return Squares[location.X][location.Y];
		}

	}
}

File: spring2013/codeWar/LRS/Clients/PlayerCppAI/GameMap.cpp
#include <cstdlib>
#include <string>
#include <sstream>
#include "GameMap.h"

using std::string;
using std::stringstream;

namespace PlayerCSharpAI
{
	namespace api
	{

		GameMap::GameMap(const xml_node& element) : m_flags(NULL)
		{
			int width = element.attribute("width").as_int();
			int height = element.attribute("height").as_int();
			m_squares = new vector< vector<MapSquare> >(width);
			for (int x = 0; x < width; x++)
			{
				(*m_squares)[x] = vector<MapSquare>(height);
				for (int y = 0; y < height; y++)
				{
					stringstream out;
					out << "//square[@x=" << x << " and @y=" << y << "]";
					xml_node elemSquare = element.select_single_node(out.str().c_str()).node();
					(*m_squares)[x][y] = MapSquare(elemSquare);
				}
			}
		}

		GameMap::~GameMap()
		{
			delete m_squares;
			delete m_flags;
		}

		int GameMap::getWidth()
		{
			return m_squares->size();
		}

		int GameMap::getHeight()
		{
			return m_squares->at(0).size();
		}

		vector< vector<MapSquare> >* GameMap::getSquares()
		{
			return m_squares;
		}

		/*void GameMap::setSquares(vector<vector<MapSquare>> value)
		{
			m_squares = value;
		}*/

		vector<Point>* GameMap::getFlags()
		{
			return m_flags;
		}

		/*void GameMap::setFlags(vector<Point> value)
		{
			m_flags = value;
		}*/

		MapSquare* GameMap::GetSquare(Point location)
		{
			return &(*m_squares)[location.X][location.Y];
		}
	}
}
